### ISSUE TITLE
Newlines in markdown being converted to `<br/>`

### ISSUE THREAD

**Issue by DannyBen** [retypeapp/retype#87]
I noticed that newlines in the markdown document are converted to `<br>`.

This is not standard markdown behavior ([reference](https://daringfireball.net/projects/markdown/syntax#p) - although I know that some implementations do it like this.

GitHub, for example, uses this approach in comments and descriptions (like this one), but avoid using this in markdown documents in source code (e.g. README.md).

I believe the reason that the GitHub engineers used this approach is this:

- Comments and descriptions are written using a UI, which already has word wrapping on, and it is used by the "average user" which might not know about the "double space" feature.
- READMEs are written in a text editor and by a "developer-class user", so there is no reason not to adhere to the standard in this case.

Of course, the ability to still create an explicit `<br/>` in the standard markdown implementation is still there, by ending a sentence with two spaces.

I believe that Retype should **not** convert newlines to `<br/>`, or at least, to have an option to turn it off.

**Comment by geoffreymcgill** [comment]
Thanks for bringing this to our attention. 

Retype uses Markdig for its Markdown parsing and we have the [HardlineBreak](https://github.com/xoofx/markdig/blob/master/src/Markdig.Tests/Specs/HardlineBreakSpecs.md) extension enabled.

```
This is a paragraph
with a break inside
```
Here's the HTML generated with the `HarlineBreak` extension enabled:

```html
<p>This is a paragraph<br />
with a break inside</p>
```
and with the `HardlineBreak` extension not enabled:
```
<p>This is a paragraph</p>
<p>with a break inside</p>
```
If the two paragraphs are separated by an empty line, then each would be treated as a separate `<p>` paragraph. Sample:
```
This is a paragraph
                                <--- empty line
with a break inside
```
Expected HTML would be:
```
<p>This is a paragraph</p>
<p>with a break inside</p>
```
Reading the CommonMark spec, I do see the [Hard line breaks](https://spec.commonmark.org/0.30/#hard-line-breaks) specified with ending the paragraph with a double-space `  ` or backslack `\` which should render as a `<br />`. 

I have to admit, I didn't actually know this hard line break syntax existed in the CommonMark spec. I'm a little perplexed why this would have been originally designed like this in Markdown. The way Markdig and others handle newlines just seems so much cleaner and expected.

The trailing double-space would drive me nuts, especially because we trim line-endings on save (VS Code setting?).

But... I can see that we are not following the spec exactly and a configuration to disable (not enable) the `HardlineBreak` extension is required. 

We have discussed adding a `markdown` config section in the `retype.json` and this scenario seems like a perfect opportunity implement. 

What do you think about the following `markdown.linebreaks` config proposal?

```jsonc
{
    "markdown": {
        "linebreaks": "hardline" // or "softline", default is "hardline"
    }
}
```

**Comment by DannyBen** [comment]
Thanks for the thoughtful and speedy response.

First of all - assuming I understood correctly - I believe you have a mistake in your understanding of the spec.

> Expected HTML would be:
>
> ```
> <p>This is a paragraph</p>
> <p>with a break inside</p>
> ```

This is incorrect, and if this is indeed the case without that plugin - then the implementation is totally wrong. In markdown, only an empty line denotes "end of a block" (paragraph, list, heading, etc).

When using **one** newline, without any trailing spaces, this is considered a **regular shitespace**.

So:

```
first line
second line

# should be rendered as a single paragraph
<p>first line second line</p>

---

first line      # <-- followed by two spaces
second line

# should be rendered as with a line break
<p>first line<br/>second line</p>

---

first line
                   # <-- empty line
second line

# should be rendered as two paragraphs
<p>first line</p>
<p>second line</p>
```

The rationale behind this design is simple: Markdown was designed so that it is clean and beautiful even when viewed in plain text. So - in the same way that when we write code, we avoid long lines despite having a "text wrap" feature in the editor, markdown lets us do the same by adding a newline in the source (so it is nice and clean as plain text) and is still smart enough to group together groups of lines into one `<p>paragraph</p>`.

Of course this is not a terrible big deal, but I for one, am **always** ensuring that my lines in markdown do not exceed 80 characters.

By the way - some other book generators (like mdbook, which is my current tool of choice) are obeying this markdown standard - so converting my sites from something else to retype is becoming more labor intensive.

As for the config - sure, looks great if it can be done (and assuming `softlines` avoids creating `<br>` or a separate `<p>`).

**Comment by geoffreymcgill** [comment]
We have added a new `markdown.lineBreaks` configuration option to the `retype.json` that enables configuration of the `soft` vs `hard` line breaks. 

```jsonc
{
   "markdown": {
      "lineBreaks": "soft|hard" // default is "soft"
   }
}
```

We have also switched the default rendering to `soft` line breaks, which follows the CommonMark spec exactly. Your original Markdown samples will now work as expected where the lines can end with a double-space `  ` or backslash `\` to render a hard line break (`<br />`). 

See CommonMark [spec](https://spec.commonmark.org/0.30/#hard-line-breaks). 

### Sample 1

No trailing spaces or backslash.

#### Markdown

```md
Paragraph one
Paragraph two
```

#### Generated HTML

```html
<p>Paragraph one Paragraph two</p>
```

### Sample 2

Two paragraphs separated by an empty line renders as two fully separate paragraphs in HTML.

#### Markdown

```md
Paragraph three

Paragraph four
```

#### Generated HTML

```html
<p>Paragraph three</p>
<p>Paragraph four</p>
```

### Sample 3

Paragraph ending with a two-space characters, creates a hard line break and renders as one paragraph with a `<br />` separating the two lines.

#### Markdown

```md
Paragraph five  
Paragraph six
```

#### Generated HTML

```html
<p>Paragraph three<br />
Paragraph four</p>
```

### Sample 4

Paragraph ending with a backslach character, creates a hard line break and renders as one paragraph with a `<br />` separating the two lines.

#### Markdown

```md
Paragraph seven\
Paragraph eight
```

#### Generated HTML

```html
<p>Paragraph seven<br />
Paragraph eight</p>
```

---

If you would prefer Retype to render line breaks where a newline char is always replaced with a `<br />`, please add the following configuration to your `retype.json`:

```jsonc
{
   "markdown": {
      "lineBreaks": "hard"
   }
}
```

Again, the default value is `soft`, which follows the CommonMark [spec](https://spec.commonmark.org/0.30/#hard-line-breaks).

This feature will be included in the upcoming Retype `v1.8.0` release.

@DannyBen Thanks for bringing this issue to our attention. 🙌

**Comment by DannyBen** [comment]
Wow - super cool. Happy to see standards compliance.

BTW - since Retype is closed source, where is the best place to subscribe to get notifications for releases? Normally it would be a GitHub "releases" notification. If there isn't any - preferably a GitHub-native option - perhaps it would be a good idea to have a long standing ticket in the Discussions for this?

**Comment by geoffreymcgill** [comment]
> where is the best place to subscribe to get notifications for releases?

You should be able to just subscribe to this repo for notifications. We publish the releases to https://github.com/retypeapp/retype/releases

Hope this helps.

Okay, you found it. :+1:

**Comment by DannyBen** [comment]
Confirmed as fixed in version `1.7.1-202107290616`
